// denamed/dist/core/answers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createAaaaAnswer = exports.createCnameAnswer = exports.createTxtAnswer = exports.createNSAnswer = exports.createMxAnswer = exports.createAAnswer = void 0;
const protocol_1 = require("./protocol");
function createAAnswer(question, ip, ttl = 60) {
    return {
        type: "A",
        ttl: ttl,
        name: question.name,
        class: protocol_1.DnsClass.Internet,
        data: {
            type: "A",
            target: ip,
        },
    };
}
exports.createAAnswer = createAAnswer;
function createMxAnswer(question, exchange, priority = 0, ttl = 60) {
    return {
        type: "MX",
        ttl: ttl,
        name: question.name,
        class: protocol_1.DnsClass.Internet,
        data: {
            type: "MX",
            exchange,
            priority
        },
    };
}
exports.createMxAnswer = createMxAnswer;
function createNSAnswer(question, target, ttl = 60) {
    return {
        type: "NS",
        ttl: ttl,
        name: question.name,
        class: protocol_1.DnsClass.Internet,
        data: {
            type: "NS",
            target
        },
    };
}
exports.createNSAnswer = createNSAnswer;
function createTxtAnswer(question, target, ttl = 60) {
    return {
        type: "TXT",
        ttl: ttl,
        name: question.name,
        class: protocol_1.DnsClass.Internet,
        data: {
            type: "TXT",
            target
        },
    };
}
exports.createTxtAnswer = createTxtAnswer;
function createCnameAnswer(question, target, ttl = 60) {
    return {
        type: "CNAME",
        ttl: ttl,
        name: question.name,
        class: protocol_1.DnsClass.Internet,
        data: {
            type: "CNAME",
            target
        },
    };
}
exports.createCnameAnswer = createCnameAnswer;
function createAaaaAnswer(question, target, ttl = 60) {
    return {
        type: "AAAA",
        ttl: ttl,
        name: question.name,
        class: protocol_1.DnsClass.Internet,
        data: {
            type: "AAAA",
            target,
            ttl: 1000
        },
    };
}
exports.createAaaaAnswer = createAaaaAnswer;
