// denamed/dist/records.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createSrvRecord = exports.createSoaRecord = exports.createTxtRecord = exports.createNsRecord = exports.createMxRecord = exports.createCnameRecord = exports.createAaaaRecord = exports.createARecord = void 0;
const createARecord = (target) => ({
    type: "A",
    target,
});
exports.createARecord = createARecord;
const createAaaaRecord = (target) => ({
    type: "AAAA",
    target,
});
exports.createAaaaRecord = createAaaaRecord;
const createCnameRecord = (target) => ({
    type: "CNAME",
    target,
});
exports.createCnameRecord = createCnameRecord;
const createMxRecord = (record) => ({
    type: "MX",
    priority: record.priority ?? 0,
    exchange: record.exchange,
});
exports.createMxRecord = createMxRecord;
const createNsRecord = (target) => ({
    type: "NS",
    target,
});
exports.createNsRecord = createNsRecord;
const createTxtRecord = (target) => ({
    type: "TXT",
    target,
});
exports.createTxtRecord = createTxtRecord;
const createSoaRecord = (record) => ({
    type: "SOA",
    admin: record.admin ?? "hostmaster." + record.host,
    host: record.host,
    serial: record.serial ?? 0,
    refresh: record.refresh ?? 10,
    retry: record.retry ?? 10,
    expire: record.expire ?? 10,
    ttl: record.ttl ?? 10,
});
exports.createSoaRecord = createSoaRecord;
const createSrvRecord = (record) => ({
    type: "SRV",
    priority: 0,
    weight: 10,
    target: record.target,
    port: record.port,
});
exports.createSrvRecord = createSrvRecord;
